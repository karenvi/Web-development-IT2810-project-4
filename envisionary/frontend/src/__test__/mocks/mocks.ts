
import { ADD_REVIEW } from "../../graphql/mutations";
import { GET_COUNTRIES, GET_REVIEWS_BY_COUNTRY_NAME } from "../../graphql/queries";





export const mocks = [
    {
      request: {
        query: GET_COUNTRIES,
        variables: {}
      },
      result: {
        data: {
            country: 
            [{ _id: "1", Area: "120538", CCA3: "PRK", Capital: "Pyongyang", Continent: "Asia", Country: "North Korea", Density: "216.2755", 
            GrowthRate: "1.0038", Population1970: "14996879", Population1980: "17973650", Population1990: "20799523", Population2000: "23367059",
            Population2010: "24686435", Population2015: "25258015", Population2020: "25867467", Population2022: "26069416", Rank: "56", 
            WorldPopulationPercentage: "0.33" },

            { _id: "2", Area: "25713", CCA3: "MKD", Capital: "Skopje", Continent: "Europe", Country: "North Macedonia", Density: "81.4218", GrowthRate: "0.9954",
            Population1970: "1656783", Population1980: "1907023", Population1990: "2044174", Population2000: "2037936", Population2010: "2093828",
            Population2015: "2107962", Population2020: "2111072", Population2022: "2093599", Rank: "150", WorldPopulationPercentage: "0.03"},

            { _id: "3", Area: "464", CCA3: "NFK", Capital: "Saipan", Continent: "Oceania", Country: "Northern Mariana Islands", Density: "106.7909", 
            GrowthRate: "1.0014", Population1970: "10143", Population1980: "17613", Population1990: "48002", Population2000: "80338", Population2010: "54087",
            Population2015: "51514", Population2020: "49587", Population2022: "49551", Rank: "210", WorldPopulationPercentage: "0"},
            ]
        }
      }
    },
    {
      request: {
        query: GET_REVIEWS_BY_COUNTRY_NAME,
        variables: { country: "Afghanistan" }
      },
      result: { 
        data: {
          countryByName: {
            Reviews: [
              { Name: "Trond", ReviewText: "Flott land 책 feriere i." , Date: "2022-10-26T13:50:35.783Z", Rating: 3.5 },
              { Name: "Hans og Grethe", ReviewText: "Fint og fredelig land for et idyllisk opphold p책 all-inclusive.", Date: "2022-10-26T13:53:34.973Z",  Rating: 5 },
              { Name: "Lisbeth", ReviewText: "Synes det var litt skummelt i Afghanistan.", Date: "2022-10-26T19:51:35.080Z", Rating: 3 }
            ]
          }
        }
      }  
    },
    {
      request: {
        query: ADD_REVIEW,
        variables: { country: "Afghanistan", reviewText: "Lorem ipsum dolor sit amet", date: "2022-10-26T13:53:34.973Z", rating: 3.5 }
      },
      restult: {
        data: {
          countryByName: {
            Reviews: [
              { Name: "Trond", ReviewText: "Flott land 책 feriere i." , Date: "2022-10-26T13:50:35.783Z", Rating: 3.5 },
              { Name: "Hans og Grethe", ReviewText: "Fint og fredelig land for et idyllisk opphold p책 all-inclusive.", Date: "2022-10-26T13:53:34.973Z",  Rating: 5 },
              { Name: "Lisbeth", ReviewText: "Synes det var litt skummelt i Afghanistan.", Date: "2022-10-26T19:51:35.080Z", Rating: 3 }
            ]
          }
        }
      }
    }
  ];


export const failMock = [
  {
    request: {
        query: GET_REVIEWS_BY_COUNTRY_NAME,
        variables: { country: "Afghanistan" }
    },
    error: new Error()
  }
];

export const noReviewsMock = [
  {
    request: {
      query: GET_REVIEWS_BY_COUNTRY_NAME,
      variables: { country: "Afghanistan" }
    },
    result: { 
      data: {
        countryByName: { Reviews: null }
      }
    }  
  }
];