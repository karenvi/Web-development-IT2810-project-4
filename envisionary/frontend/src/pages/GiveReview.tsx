import '../styles/App.css'
import { useLocation } from 'react-router-dom';
import { Alert, Button, IconButton, Modal, Rating, Snackbar, TextField, Typography } from '@mui/material';
import EditIcon from '@mui/icons-material/Edit';
import StarIcon from '@mui/icons-material/Star';
import { Box } from '@mui/system';
import { useContext, useState } from 'react';
import { useMutation } from '@apollo/client';
import { ADD_REVIEW } from "../graphql/mutations"
import { GET_COUNTRY_DATA_BY_NAME } from '../graphql/queries';
import CloseIcon from '@mui/icons-material/Close';
import { ThemeContext } from '../App';
import { useRecoilState } from 'recoil';
import { starOpacityRating, toggleColorTheme } from '../states/states';
import { buttonStyling } from '../components/Countries';

// Miscellanous styling both for light and dark theme
const styleTitleOfReviews = { width: "100%", display: 'flex', justifyContent: 'flex-start', mb: "20px"};

// Styling of modal
const modalStyle = {
  position: 'absolute' as 'absolute',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)',
  width: '50vw',
  maxWidth: '500px',
  minWidth: '100px',
  bgcolor: '#ffffff',
  p: 4,
  display: 'flex', 
  flexDirection: 'column', 
  alignItems: 'flex-start',
  boxShadow: '0px 2px 1px -1px rgb(0 0 0 / 20%), 0px 1px 1px 0px rgb(0 0 0 / 14%), 0px 1px 3px 0px rgb(0 0 0 / 12%)',
  borderRadius: '10px',
};

const reviewHeaderStyling = { mt: 3, fontSize: '18px' }

// Dark and light mode styling
const giveReviewStyle = {
  dark: {
      backgroundColor: '#1e374c',
      color: '#ffffff',
  },
  light: {
      backgroundColor: 'white',
      color: 'black',
  },
}

const inputReviewStyle = {
    dark: {
      width: '100%',
      input: {
        color: "#ffffff",
      },
      '& fieldset.MuiOutlinedInput-notchedOutline': {
        borderColor: '#ffffff',
      },
      '&:hover fieldset.MuiOutlinedInput-notchedOutline': {
        borderColor: '#ffffff',
      },
    },
    light: {
      width: '100%',
    },  
}


function GiveReview() {
  const { theme } = useContext(ThemeContext);
  const location = useLocation();
  const [rating, setRating] = useState<number>(0);
  const [author, setAuthor] = useState('');
  const [reviewText, setReviewText] = useState('');
  const [invalidAuthor, setInvalidAuthor] = useState<boolean>(false);
  const [authorError, setAuthorError] = useState<string>(" ");
  const [clear, setClear] = useState("false");
  const [open, setOpen] = useState(false);
  const [openModal, setOpenModal] = useState(false)
  const [starOpacity] = useRecoilState<number>(starOpacityRating);
  const [toggleColor] = useRecoilState<string>(toggleColorTheme);

  //Functions of setters that are being used multiple times to clear the review field:
  const clearReview = () => {
    setAuthor("");
    setReviewText("");
    setRating(0);
    setOpenModal(false);
  }
  // Use ADD_REVIEW mutation to add review to database
  const [addReview] = useMutation(ADD_REVIEW, {
    refetchQueries: [ // keep local cache updated by refetching reviews after adding new review 
      {query: GET_COUNTRY_DATA_BY_NAME, variables: {country: location.state.country.Country}},
    ]});

  // onChange-functions for the modal that opens "Give review"
  const handleModalOpen = () => setOpenModal(true);
  const handleModalClose = () => {
    setOpenModal(false);
    clearReview();
  }


  // Validation of give review input fields 

  const validation = () => {
    const validateNameRegex = /^[a-zA-Z]/; // names should start with normal letters
    const emptyFieldRegex = /^\s*$/; // checks for whitespaces

    if (author.length > 40) {
      setInvalidAuthor(true);
      setAuthorError("Name cannot be longer than 40 characters");
      return false
    }
    else if (emptyFieldRegex.test(author)) { // if name field is empty
      setInvalidAuthor(true);
      setAuthorError("Name is required");
      return false
    }
    else if (!validateNameRegex.test(author)) { // if name does not start with normal letters
      setInvalidAuthor(true);
      setAuthorError("A name must start with normal letters (a-z)");
      return false
    }
    else { // else, name is valid
      setInvalidAuthor(false);
      setAuthorError(" ");
      return true
    }
  }

  const submit = () => {
    if (validation()) {
      addReview({
        variables:
        {
          country: location.state.country.Country,
          name: author,
          reviewText: reviewText,
          date: new Date().toISOString(),
          rating: rating
        }
      });
      setOpen(true);// Opens the success alert.
      clearReview();

      // Clears the country field
      if (clear === "false") {
        setClear("true")
      } else {
        setClear("false")
      }
    }
  }

  const handleClose = (event?: React.SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }
    setOpen(false);
  }



  const reviewStyle = {
    ...(theme === 'light' ? giveReviewStyle.light : giveReviewStyle.dark),
  }

  const reviewInputStyle = {
    ...(theme === 'light' ? inputReviewStyle.light : inputReviewStyle.dark),
  }

  return (
    <>
    <Box sx={styleTitleOfReviews}><Typography component="h2" variant="h6">Reviews of {location.state.country.Country}:</Typography></Box>
    <Box sx={styleTitleOfReviews}>
      <Button
        endIcon={<EditIcon />}
        id="button-write-review"
        variant="contained"
        onClick={handleModalOpen}
        sx={buttonStyling}
      >
       Review {location.state.country.Country}
      </Button>
      </Box>
      <Modal open={openModal}
        onClose={handleModalClose}
      >
        
      <Box sx={modalStyle} style={reviewStyle}>
        <Box sx={{display: 'flex', justifyContent: 'space-between', width: '100%'}}>
          <Typography component="h1" variant="h6">Write a review for {location.state.country.Country}</Typography>
          <IconButton aria-label="close modal" onClick={handleModalClose}>
            <CloseIcon />
          </IconButton>
        </Box>
        <Typography component="label" htmlFor="name-field" variant="h6" sx={{ mt: 1, fontSize: '18px' }}>Name *</Typography>
        <Box sx={{width: '50%', minWidth: '100px'}}>
        <TextField id="name-field"
          required
          label=""
          placeholder="Name"
          variant="outlined"
          value={author}
          error={invalidAuthor}
          helperText={authorError}
          onChange={(e) => setAuthor(e.target.value)}
          sx={reviewInputStyle}
        />
        </Box>

        <Typography component="label" htmlFor="rating-stars" variant="h6" sx={{ mt: 1, fontSize: '18px' }}>Rating</Typography>
        <Rating id="rating-stars"
          name="hover-feedback"
          value={rating}
          precision={0.5}
          onChange={(event, newValue) => {
            newValue === null ? setRating(0) : setRating(newValue);
          }}
          emptyIcon={<StarIcon style={{ opacity: starOpacity }} fontSize="inherit" />}
        />

        <Typography component="label" htmlFor="review-content-field" variant="h6" sx={reviewHeaderStyling}>Review Content</Typography>
        <Box sx={{width: '50vw', minWidth: '100px', maxWidth: '450px', mb: "20px"}}>
        <TextField
          id="review-content-field"
          label=""
          placeholder="Write your review..."
          multiline
          rows={7}
          value={reviewText}
          onChange={(e) => setReviewText(e.target.value)}
          sx={reviewInputStyle}
          InputProps={{ inputProps: { style: { color: toggleColor }}}}
        />
        </Box>

        <Button variant="contained"
          sx={buttonStyling}
          onClick={(event) => {
            event.preventDefault();
            submit()
          }}
        >
          Submit
        </Button>
      </Box>
      </Modal>
      <Snackbar open={open} autoHideDuration={6000} onClose={handleClose}>
        <Alert onClose={handleClose} severity="success" sx={{ width: '100%' }}>
          Review successfully given!
        </Alert>
      </Snackbar>
    </>
  );
}

export default GiveReview;
