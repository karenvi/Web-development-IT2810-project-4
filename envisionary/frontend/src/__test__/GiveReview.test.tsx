import GiveReview from "../pages/GiveReview";
import { render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { mocks } from './mocks/mocks';
import userEvent from "@testing-library/user-event";


/**
 * Mocking useLocation in order to make the query mocks work for useQuery in component.
 * Technique found at https://blog.changani.me/2022/01/30/webdev/react/react-mocking-react-router-dom-s-use-location-with-jest/
 */ 
jest.mock('react-router-dom', () => {
    return {
        ...jest.requireActual('react-router-dom'),
        useLocation: () => {
            return {
                pathname: "/country",
                search: "",
                hash: "",
                state: {
                    country: {
                        Country: "Afghanistan",
                    }
                },
                key: "eb77n1u7"
            }
        }
    };
});

describe("Testing review content", () => {

    beforeEach(async () => {
        render(
            <MockedProvider mocks={mocks} addTypename={false}>
                    <GiveReview />
            </MockedProvider>
        );
    });

    it("snapshot test", () => {

        const component: renderer.ReactTestRenderer = renderer.create(
                <MockedProvider mocks={mocks} addTypename={false}>
                    <GiveReview />
                </MockedProvider>
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot(); 
    });


    it("empty review content field", async () => {

        userEvent.click(await waitFor(() => screen.findByText('Review Afghanistan')));
        expect(await screen.findByTestId('review-content-field-test')).toHaveValue('');
    });


/**
 * Couldn't get this to work. There's something different about the contend field compared to the name field
 * that makes some async stuff go wrong. Jest insists that things should be wrapped in act(), but that's a dead end. 
 *  */    

/* 
    it("entering review content", async () => {
        render(
            <MockedProvider mocks={mocks} addTypename={false} defaultOptions={{watchQuery: { fetchPolicy: 'no-cache' }, query: { fetchPolicy: 'no-cache' }, }}>
                    <GiveReview />
            </MockedProvider>
        );

        userEvent.click(await waitFor(() => screen.findByText('Review Afghanistan')));

        // content field should be empty first
        await waitFor(() => expect(screen.getByText('Review Afghanistan')).not.toBeInTheDocument());
        
        // should contain name after typing
        userEvent.type(screen.getByTestId('review-content-field-test'), "Muahaha, if it isn't Sonic >:) ");
        expect(await screen.findByTestId('name-field-test')).toHaveValue("Muahaha, if it isn't Sonic >:) ");
    });
 */
    it("empty name field", async () => {

        userEvent.click(await waitFor(() => screen.findByText('Review Afghanistan')));

        // empty name field should trigger error
        expect(await screen.findByTestId('name-field-test')).toHaveValue('');
        userEvent.click(await screen.findByText('Submit'));
        expect(await screen.findByText('Name is required')).toBeInTheDocument();
    });

    it("too long name", async () => {

        userEvent.click(await waitFor(() => screen.findByText('Review Afghanistan')));

        // name > 40 should trigger error
        userEvent.type(await screen.findByTestId('name-field-test'), '{selectall}{backspace}');
        userEvent.type(await screen.findByTestId('name-field-test'), 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam sit amet porttitor dui. Ut vel est.');
        expect(await screen.findByTestId('name-field-test')).toHaveValue('Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam sit amet porttitor dui. Ut vel est.');
        userEvent.click(await screen.findByText('Submit'));
        expect(await screen.findByText('Name cannot be longer than 40 characters')).toBeInTheDocument();
    });

    it("entering valid name", async () => {

        userEvent.click(await waitFor(() => screen.findByText('Review Afghanistan')));

        // name field should be empty first
        expect(await screen.findByTestId('name-field-test')).toHaveValue('');
        // should contain name after typing
        userEvent.type(await screen.findByTestId('name-field-test'), 'Doctor Ivo Robotnik');
        expect(await screen.findByTestId('name-field-test')).toHaveValue('Doctor Ivo Robotnik');
        // normal name should pass
        userEvent.click(await screen.findByText('Submit'));
        expect(await screen.findByText('Review successfully given!')).toBeInTheDocument();  
    });

    it("special characters in name", async() => {

        userEvent.click(await waitFor(() => screen.findByText('Review Afghanistan')));
        // name starting with special characters should trigger error
        userEvent.type(await screen.findByTestId('name-field-test'), '{selectall}{backspace}');
        userEvent.type(await screen.findByTestId('name-field-test'), '$/2gutta');
        expect(await screen.findByTestId('name-field-test')).toHaveValue('$/2gutta');
        userEvent.click(await screen.findByText('Submit'));
        expect(await screen.findByText('A name must start with normal letters (a-z)')).toBeInTheDocument();
    });
});