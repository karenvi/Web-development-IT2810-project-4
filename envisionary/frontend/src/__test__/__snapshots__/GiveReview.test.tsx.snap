// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GiveReview component snapshot test 1`] = `
Array [
  <div
    className="MuiBox-root css-jhhoj"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
    >
      Reviews of 
      Afghanistan
      :
    </h2>
  </div>,
  <div
    className="MuiBox-root css-jhhoj"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-142aol9-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      id="button-write-review"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Review 
      Afghanistan
      <span
        className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="EditIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </span>
    </button>
  </div>,
]
`;
